[bgfx] Update to v1.118.8398-400

bwrsandman opened the pull request:
**Describe the pull request**

Supersedes #28664, fixes #29029.

This PR fixes regressions introduced in #28377 where `libsquish`, `tinyexr` and `nvtt` were removed from 3rd party of bgfx and made to use packages in vcpkg. The changes were not tested with the `bgfx[tools]` features and the android platform which was supported previously wasn't tested either.

The biggest issue is that [bgfx's 3rd party nvtt](https://github.com/bkaradzic/bimg/blob/f0ba7cb88cdd13651ca3691230c723e0c8090c45/3rdparty/nvtt/nvtt.cpp) does not match [what's in the `nvtt` package](https://github.com/castano/nvidia-texture-tools/blob/e393b6db0483f1fc41d85a6da18ec961e83ccd82/src/bc6h/zoh.cpp). They don't provide the same functions. Bgfx's is completely stripped down and with many fixes.

The #28377 PR broke a few things:

1. Installing on android
>    Error: nvtt:arm-android@2.1.2#8 is only supported on '!uwp & !arm'

2. Configuring projects
```
CMake Error at cmake-build-presets/ninja-multi-vcpkg/vcpkg_installed/x64-linux/share/bgfx/bgfxTargets.cmake:69 (set_target_properties):
  The link interface of target "bgfx::bimg" contains:

    unofficial::libsquish::squish

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  cmake-build-presets/ninja-multi-vcpkg/vcpkg_installed/x64-linux/share/bgfx/bgfxConfig.cmake:27 (include)
  vcpkg/scripts/buildsystems/vcpkg.cmake:843 (_find_package)
  CMakeLists.txt:87 (find_package)

CMake Error at cmake-build-presets/ninja-multi-vcpkg/vcpkg_installed/x64-linux/share/bgfx/bgfxTargets.cmake:69 (set_target_properties):
  The link interface of target "bgfx::bimg" contains:

    unofficial::tinyexr::tinyexr

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  cmake-build-presets/ninja-multi-vcpkg/vcpkg_installed/x64-linux/share/bgfx/bgfxConfig.cmake:27 (include)
  vcpkg/scripts/buildsystems/vcpkg.cmake:843 (_find_package)
  CMakeLists.txt:87 (find_package)

```

4. Compiling with the `tools` feature
```
FAILED: texturec 
: && /usr/bin/c++ -fPIC -g -rdynamic vcpkg/buildtrees/bgfx/src/25d5e8d3da-47d8177139.clean/tools/texturec/texturec.cpp.o -o texturec  libbimg.a  libastc-encoder.a  libedtaa3.a  libetc1.a  libetc2.a  libbx.a  -lrt  libiqa.a  vcpkg_installed/x64-linux/debug/lib/libsquishd.a  vcpkg_installed/x64-linux/lib/static/libnvtt.a  libpvrtc.a  vcpkg_installed/x64-linux/share/tinyexr/../../debug/lib/libtinyexr.a  -ldl && :
/usr/bin/ld: libbimg.a(image_encode.cpp.o): in function `bimg::imageEncodeFromRgba32f(bx::AllocatorI*, void*, void const*, unsigned int, unsigned int, unsigned int, bimg::TextureFormat::Enum, bimg::Quality::Enum, bx::Error*)':
vcpkg/buildtrees/bgfx/src/25d5e8d3da-47d8177139.clean/src/image_encode.cpp:272: undefined reference to `nvtt::compressBC6H(void const*, unsigned int, unsigned int, unsigned int, void*)'
```

- #### What does your PR fix?
* Update bgfx.cmake to v1.118.8398-400 which no longer produces 3rd party libraries.
* Removes dependency on vcpkg's `nvtt` which does not match what bgfx uses.
* Add custom user script to inject vcpkg packages for `miniz`, `tinyexr` and `libsquish` instead of using patches.
* Fixes compiling android.
* Fixes packages missing when linking with the `bgfx` package on all platforms.
* Fixes compiling the `tools` feature.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
all

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
Yes

JonLiu1993 commented:
All features are tested successfully in the following triplet:
- x86-windows
- x64-windows
- x64-windows-static

bwrsandman commented:
Requires a fix from #171 so putting it in draft until then,

dg0yt commented:
> * Removes dependency on vcpkg's `libsquish`, `tinyexr` and `nvtt`.

This restores the vendored copies. This is not how vcpkg works. I see the problem with the unmaintained nvtt. But why not use properly de-vendor the other libraries?

bwrsandman commented:
The author (@bkaradzic) specifically doesn't want to use 3rd parties that he doesn't have source control over. 

1. Modifications like those in nvtt happen in other 3rd parties
2. He uses specific versions of these 3rd parties and worries bugs might come from 3rd parties introducing bugs or changing behavior and not being able to reproduce using his own source.

Doing things this way aren't my personal preference, I would love to use vcpkg's spriv-tools for example but I'm trying to respect the author's wishes while at the same time making bgfx work in the cmake and vcpkg world.

bkaradzic commented:
@bwrsandman In official repo those are not even separate libraries, rather they are part of bimg_encode library:
https://github.com/bkaradzic/bimg/blob/1395f4e969fa0aac6158fb3caf0873eaed38d77f/scripts/bimg_encode.lua#L6-L37

Probably would be better fix to make CMake to match official repo.

bwrsandman commented:
The following patch works for using the package libs, but matching the official repo is probably the best thing to do. 

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1b0a893..9e056ec 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -64,6 +64,11 @@ if( BGFX_LIBRARY_TYPE MATCHES "SHARED" )
     set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 endif()
 
+find_package(unofficial-libsquish CONFIG REQUIRED)^M
+add_library(squish ALIAS unofficial::libsquish::squish)^M
+find_package(tinyexr CONFIG REQUIRED)^M
+add_library(tinyexr ALIAS unofficial::tinyexr::tinyexr)^M
+^M
 if( NOT BX_DIR )
        set( BX_DIR "${CMAKE_CURRENT_SOURCE_DIR}/bx" CACHE STRING "Location of bx." )
 elseif( NOT IS_ABSOLUTE "${BX_DIR}")
@@ -149,7 +154,7 @@ if( BGFX_INSTALL )
                         INCLUDES DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
        )
        if( NOT BGFX_LIBRARY_TYPE MATCHES "SHARED" )
-               install( TARGETS bimg bx astc-encoder edtaa3 etc1 etc2 iqa squish nvtt pvrtc tinyexr
+               install( TARGETS bimg bx astc-encoder edtaa3 etc1 etc2 iqa nvtt pvrtc^M
                                 EXPORT "${TARGETS_EXPORT_NAME}"
                                 LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
                                 ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
```

vicroms commented:
We de-vendored the dependencies on `bgfx` in the first place because they were causing issues with the rest of our ports (see baseline regressions fixed by #28377). So going back to vendoring copies of these dependencies is not an option. 

For `nvtt` specifically we see 3 possible options:
1. Namespace all symbols in the vendored copy of `nvtt` to prevent collisions with vcpkg's `nvtt`. Since `bgfx`'s copy of `nvtt` is already heavily modified this might be acceptable. 
2. Patch `bgfx` to work with the copy of `nvtt` in vcpkg (this might be a lot of work).
3. Delist `bgfx` from the main catalogue and put it in a custom registry (https://github.com/microsoft/vcpkg/discussions/17161)

vicroms reviewed (changes requested):
Do not restore vendored copies of dependencies.

bwrsandman commented:
Working with the author to remove vendored copy of nvtt, squish, tinyexr and others. Will have something in the coming days / weeks.

bwrsandman committed:
[bgfx] Update to v1.118.8398-400

Remove vendored copies of 3rdparty bimg dependencies.
No more installed 3rdparty static libraries from bimg.
Add `BGFX_CMAKE_USER_SCRIPT` as entry point to change functionality of
bgfx.cmake config without the need for patches.
Insert miniz, tinyexr and libsquish from vcpkg repo into config.
Remove vcpkg's nvtt because bgfx's is completely different.

bwrsandman commented:
@vicroms the latest update removes the vendored copies of the dependencies and uses bgfx's packages. On upstream, `nvtt` is no longer compiled as a library. There is no need for aliasing because the `nvtt` symbols in bgfx are `compressBC6H` and `compressBC7`. These symbols are not in `vcpkg`'s nvtt package.

bwrsandman commented on ports/bgfx/vcpkg_dependencies.cmake:
```diff
@@ -0,0 +1,2 @@
+find_package(miniz CONFIG REQUIRED)
```
Shouldn't be there

JonLiu1993 committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into bgfx-v1.118.8398-365

BillyONeal commented:
Thank you for updating and avoiding revendoring the libraries that caused conflicts before!
